quickwit-storage/src/debouncer.rs

```diff
@@ -0,0 +1,307 @@
+// Copyright (C) 2022 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::hash::Hash;
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+
+use async_trait::async_trait;
+use fnv::FnvHashMap;
+use futures::future::{BoxFuture, Shared};
+use futures::{Future, FutureExt};
+use tantivy::directory::OwnedBytes;
+
+use crate::{Storage, StorageResult};
+
+/// The AsyncDebouncer debounces inflight Futures, so that concurrent async request to the same data
+/// source can be deduplicated.
+///
+/// Since we pass the Future potentially to multiple consumer, everything needs to be cloneable. The
+/// data and the future. This is reflected on the generic type bounds for the value V: Clone.
+///
+/// Since most Futures return an Result<V, Error>, this also encompasses the error.
+pub struct AsyncDebouncer<K, V: Clone> {
+    cache: Mutex<FnvHashMap<K, Shared<BoxFuture<'static, V>>>>,
+}
+
+impl<K, V: Clone> Default for AsyncDebouncer<K, V> {
+    fn default() -> Self {
+        Self {
+            cache: Default::default(),
+        }
+    }
+}
+
+impl<K: Hash + Eq + Clone, V: Clone> AsyncDebouncer<K, V> {
+    /// Instead of the future directly, a constructor to build the future is passed.
+    /// In case there is already an existing Future for the passed key, the constructor is not
+    /// used.
+    pub async fn get_or_create<T, F>(&self, key: K, build_a_future: T) -> V
+    where
+        T: FnOnce() -> F,
+        F: Future<Output = V> + Send + 'static,
+    {
+        // explicit scope to drop the lock
+        let fut_opt = { self.cache.lock().unwrap().get(&key).cloned() };
+        if let Some(future) = fut_opt {
+            return future.await;
+        }
+
+        let fut = Box::pin(build_a_future()) as BoxFuture<'static, V>;
+        let fut = fut.shared();
+        self.cache.lock().unwrap().insert(key.clone(), fut.clone());
+        let res = fut.await;
+
+        self.cache.lock().unwrap().remove(&key);
```

Comment:
I am a bit scared of the possibity for a consumer to call 
`get_or_create` and the dropping the resulting future.

The unbouncing map would be populated, but the future would then not be polled for a while. I then another consume comes in an asks for the future, it could get an error, not because the value is wrong but because of a server timeout for instance.

Reply:
(Note that I am not sure what we should do here)

Reply:
I think the future always gets polled to completion since we clone and  await the future in `get_or_create`